test/gtest/uct/ib/test_sockaddr.cc

```diff
@@ -852,6 +870,30 @@ UCS_TEST_P(test_uct_cm_sockaddr, connect_client_to_server_reject_with_delay)
     test_delayed_server_response(true);
 }
 
+UCS_TEST_P(test_uct_cm_sockaddr, double_ep_disconnect)
+{
+    ucs_status_t status;
+
+    cm_listen_and_connect();
+
+    wait_for_bits(&m_cm_state, TEST_CM_STATE_SERVER_CONNECTED |
+                               TEST_CM_STATE_CLIENT_CONNECTED);
+    EXPECT_TRUE(ucs_test_all_flags(m_cm_state, (TEST_CM_STATE_SERVER_CONNECTED |
+                                                TEST_CM_STATE_CLIENT_CONNECTED)));
+
+    ASSERT_UCS_OK(uct_ep_disconnect(m_client->ep(0), 0));
+
+    /* wrap errors since the client will call uct_ep_disconnect the second time
+     * on the same endpoint */
+    {
+        scoped_log_handler slh(detect_double_disconncet_error_logger);
+
+        status = uct_ep_disconnect(m_client->ep(0), 0);
+        EXPECT_TRUE((status == UCS_ERR_INVALID_PARAM) ||
```

Comment:
Why second call after connection was established can return BUSY?
we can check it separately:
1) you can check TEST_CM_STATE_CLIENT_CONNECTED before wait_for_bits and try to disconnect if ep is not connected, it should fail with BUSY
2) do disconnect then it's connected - result should be UCS_OK.
3) any next trying should fail with INVALID PARAM.

pls note that (1) might require lock

Reply:
Socond call after ep is connected will return invalid_param. I don't check the state of the ep before the second call so both cases are possible and tested.
I can split it to two cases or two tests with a lock but IMHO it's a bit of an overkill for just checking a return status..

Reply:
I see the first call at line 884 and it's expected to be successful because you are waiting both client and server are connected at line 881. So, here only INVALID_PARAM is possible or you have a bug in implementation.

Reply:
But if in the second disconnect the ep is still connected then busy will be returned.

Reply:
IMO it's a bug, any disconnect call after success one should return INVALID_PARAM

Reply:
pls see @yosefe 's comment above:
"if already disconnected - return some error
if started but not finished - return busy"

IMHO it's not very critical since this an error flow and a ucs_error would be printed to the user anyway.

Reply:
I now see that this comment was about the first disconnect call.
@yosefe when should BUSY be returned then? only in the first disconnect call?

Reply:
uct_ep_create()
uct_ep_disconnect() -> UCS_OK (if no other error)
uct_ep_disconnect() -> UCS_ERR_BUSY
<got disconnect callback>
uct_ep_disconnect() -> UCS_ERR_NOT_CONNECTED (add new status code)

Reply:
@yosefe why should second disconnect return BUSY?
IMO the behavior  should be like following:
in case of success:
uct_ep_create()
uct_ep_disconnect() -> UCS_ERR_BUSY
(got connect callback with UCS_OK status)
uct_ep_disconnect() -> UCS_OK
uct_ep_disconnect() -> UCS_ERR_INVALID_PARAM
(got disconnect callback)
uct_ep_disconnect() -> UCS_ERR_INVALID_PARAM
uct_ep_destroy()

in case of failed connect:
uct_ep_create()
uct_ep_disconnect() -> UCS_ERR_BUSY
(got connect callback with UCS_ERR_*)
uct_ep_disconnect() -> UCS_ERR_INVALID_PARAM (it's not connected)
uct_ep_destroy()

Reply:
We should not use INVALID_PARAM as a generic error to say "you are doing something wrong"
In this case there is no real invalid parameter (ep is valid). If ep is not connected yet we should return ERR_NOT_CONNECTED. If disconnect already started we return BUSY to mean the process is already underway.